### PR TITLE
Remove Relative Rank on Pinned Pawns

### PR DESCRIPTION
**Pull request by miguel-l**
Previously we use half of the ThreatByRank bonus for pinned opponent
pawns. Instead, we can do a simplification by collecting all relevant
threats and use a bitwise AND with opponent pawns and do a popcount.

Since the rank of the pawns turned out to be irrelevant, I have also
named the bonus differently to conceptually reflect this, as well as to
make future tweaks easier.

Aside from conceptual simplicity, this also:
* potentially removes conditional checking for blockers in two while loops
* removes additional calls to relative_rank
* potentially performs less multiplications

Of course the trade-off is some cheap bit manipulations and a single popcount.

Also tried was a simpler version which removes the bitwise AND with
opponent pawns (thereby considering all enemy pieces) but it failed STC,
perhaps some tweaks can change this.

STC:
LLR: 2.95 (-2.94,2.94) [-3.00,1.00]
Total: 21067 W: 4797 L: 4674 D: 11596
http://tests.stockfishchess.org/tests/view/5b5af7760ebc5902bdb8d3bb

LTC:
LLR: 2.96 (-2.94,2.94) [-3.00,1.00]
Total: 48638 W: 8288 L: 8212 D: 32138
http://tests.stockfishchess.org/tests/view/5b5b001e0ebc5902bdb8d6a4

Bench: 4713919

### CODE REVIEW

**Comment by snicolet** [comment]
Hmm, to me this is not clear that it qualifies as a simplification, for the following reasons:

• I agree that that there is a conceptual simplification with the absence of the rank condition
• however, current master gives the bonus for pinned threats on all pieces, you are restricting them to pinned pawns
• current master uses two 'AND', you are replacing that by a popcount() call
•  introduction of a new score constant.

In short, this is more like replacing one algorithm by another algorithm of the same complexity. We usually test these kinds of changes with [0..5] bounds.

Of course, if you managed to prove an Elo gain by your implementation (or a variant), then I would be happy to commit :-)

**Comment by miguel-l** [comment]
@snicolet 
My thought process was as follows:
* At first I wanted to just remove relative_rank / 2 to see if it makes a difference, I think we could agree it would qualify as a simplification at that point.
* Now that leaves a weird ThreatByRank bonus without a dependency on rank, which in my view is inappropriate, as conceptually, it would not make sense. Probably better is to call it something else, hence the new score constant.
* After the above, there seems to be no need to keep it in two loops, we don't need to get the relative rank of s and its piece type is easy to get with an 'AND'. I agree it's just replacing one algorithm with another, and I probably could have done without this.

I do think there is a simplification somewhere here, it just feels to me as though doing the first step above leads naturally to the others, and I also think it's harder to change the current master to find ELO gaining patches than it would be if this were committed. Of course I'm not sure if others would agree, since it's a subjective thing. If anything, I'd be happy to retest this as a simplification without the last step, as I do see your point about changing the algorithm.

As for the current master giving bonus for pinned threats, it is also restricting the current ThreatByRank bonus to pinned pawns (an 'else if' with the previous 'if' checking for non-pawns), so I don't see much of a difference here.

**Comment by snicolet** [comment]
Sorry, I think this does not qualify as a simplification, so I reject the patch.
